# Buying and Selling - Exploring eBay Car Sales

We will be looking at a dataset about useed cars from eBay Kleinanzeigen. This is a seection of the German eBay website. The aim of this analysis is to clean the data and analyse the used car listings.


In [1]:
#impoert pandas and numpy modules
import pandas as pd
import numpy as np

In [2]:
#read in the data
autos = pd.read_csv('/Users/yasmeenmahomedy2/Downloads/Datasets/Autos.csv')

In [3]:
#inspect data
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

## Organising Columns


Not all column names are accurately descriptive, so we will rename some for better understanding. The column names are also in camelcase, not standard snakecase. To make analysis easier, we will convert to snakecase.

In [5]:
#printing names of columns for easier reference
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
#renaming columns in snakecase
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'num_pics', 'postal_code', 'last_seen']

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pics  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00         0        79588   
1         bmw              nein  2016-04-04 00:00:00         0        71034   
2  volkswagen              nein  2016-03-26 00:00:00         0        35394   
3       smart              nein  2016-03-12 00:00:00         0        33729   
4        ford              nein  2016-04-01 00:00:00         0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Initial Cleaning and Exploration

In [8]:
#using .describe and .info to see an overview of the data
autos.describe(include='all') #include null values

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-14 20:50:02  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pics   postal_code            last_seen  
count                 50000   50000.0  50000.000000                50000  
unique                   76       NaN           NaN                39481  
top     2016-04-03 00:00:00       NaN           NaN  2016-04-07 06:17:27  
freq                   1946       NaN           NaN                    8  
mean                    NaN       0.0  50813.627300                  NaN  
std                     NaN       0.0  25779.747957                  NaN  
min                     NaN       0.0   1067.000000                  NaN  
25%                     NaN       0.0  30451.000000                  NaN  
50%                     NaN       0.0  49577.000000                  NaN  
75%                     NaN       0.0  71540.000000                  NaN  
max                     NaN       0.0  99998.000000                  NaN

In [9]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  object
 5   ab_test             50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

From a glance we can see:

* Most columns are strings, so we can't find mean, max, median, ect. However, we can still see the most common entries
* In the num_pics column, most entries are 0.0. We neeed to investigate this further
* Looking at the frequency of the top values and the count, we can deduce the seller column and offer_type have mostly identical values
* The price and odometer columns are numerical values stored as text

## Closer Look at Potentially Empty Columns

In [10]:
#checking the unique values in the num_pics column
autos["num_pics"].value_counts()

0    50000
Name: num_pics, dtype: int64

In [11]:
#looking at first 10 rows of the column
autos["num_pics"].head(10)

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
9    0
Name: num_pics, dtype: int64

It seems like the num_pics column is composed of all zero values, so we will drop this column.


In [12]:
#checking the sellerr column values
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [13]:
#checking offer_type column values
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

From the above we can see the seller and offer_type columns are also composed almost entirely of the same entries. So we will drop them as well.

In [14]:
#removing num_pics, seller, and offer_type columns
autos = autos.drop(["num_pics", "seller", "offer_type"], axis=1)

## Converting String columns to Numerical Columns

The price and odometer columns are currently in string format. To make our analysis easier we will clean the columns then convert to integer values.

In [15]:
#check values in price column
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [16]:
#remove $ from price column, and extra punctuation
autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int))
autos["price"].head()

<ipython-input-16-fee836861a67>:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  autos["price"] = (autos["price"]


0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [17]:
#remove km from odometer column, and extra punctuation 
autos["odometer"] = (autos["odometer"]
                             .str.replace("km","")
                             .str.replace(",","")
                             .astype(int)
                             )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

## Inspect Price and Odometer columns:

Let's check the statistics of these columns to make sure there are no outliers or incorrect values.

In [18]:
#odometer exploration
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [19]:
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())
print(autos["odometer_km"].value_counts(ascending=True))

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
10000       264
20000       784
30000       789
40000       819
5000        967
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64


From the odometer column we can see more entries have higher values. The mean is approximately 125733km. 150000km is the most common distance. All the values are rounded integers, which is unusual for all the cars to fit into these categories. This suggests there are preset values and the milage of the cars is sorted into the closest category.

In [20]:
autos["price"].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
6770        1
61999       1
20987       1
6578        1
48600       1
Name: price, Length: 2357, dtype: int64

In [21]:
#price exploration
print(autos["price"].unique().shape)
print('\n')
print(autos["price"].describe())
print('\n')
print(autos["price"].value_counts().head(15))


(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
Name: price, dtype: int64


The prices also seem rounded off, like the odometer column. There are 2357 unique values in the column, out of a total of 50000. We need to inspect this further to see if it's worth removing this column. The maxium and minimum prices seem off, so we will explore this further to see if they are outliers.


In [22]:
#view first 20 values of price column, sorted decendingly
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [23]:
#viewing price in ascending order
autos["price"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

We can see there are a lot of cars that sell for under USD 2, since e-bay is an auction site it may be legitimate that a car sells for USD 1. We will remove USD 0 entries, as this seems unlikey. About 14 listings are above USD 1 million. This amount seems unlikely for used eBay cars. We will remove the values above USD 350,000 as there is a large jump from this value to the higher ones. 

In [24]:
#remove values in the above gap
autos = autos[autos["price"].between(1,351000)]

In [25]:
#check our above selection is correct
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the Date Column

In [26]:
print(autos.columns)

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')


In [27]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48565 non-null  object
 1   name                48565 non-null  object
 2   price               48565 non-null  int64 
 3   ab_test             48565 non-null  object
 4   vehicle_type        43979 non-null  object
 5   registration_year   48565 non-null  int64 
 6   gearbox             46222 non-null  object
 7   power_ps            48565 non-null  int64 
 8   model               46107 non-null  object
 9   odometer_km         48565 non-null  int64 
 10  registration_month  48565 non-null  int64 
 11  fuel_type           44535 non-null  object
 12  brand               48565 non-null  object
 13  unrepaired_damage   39464 non-null  object
 14  ad_created          48565 non-null  object
 15  postal_code         48565 non-null  int64 
 16  last_seen           48

There are 5 columns which should represent dates: date_crawled, registation_year, registation_month, ad_created, and last_seen. 
date_crawled, ad_created, and last_seen are all stored as strings, so we need to convert them to analyse the data.

In [28]:
#looking at the data from 3 columns, only looking at first 5 entries
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

### Date Crawled Column:

In [29]:
print(autos['date_crawled'].str[:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object


In [30]:
#view first 10 rows of date_crawled column, in proportions
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False)

2016-04-03    0.038608
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-12    0.036920
2016-03-14    0.036549
2016-04-04    0.036487
2016-03-07    0.036014
2016-04-02    0.035478
2016-03-28    0.034860
2016-03-19    0.034778
2016-03-15    0.034284
2016-03-29    0.034099
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-22    0.032987
2016-03-11    0.032575
2016-03-23    0.032225
2016-03-26    0.032204
2016-03-10    0.032184
2016-03-31    0.031834
2016-03-17    0.031628
2016-03-25    0.031607
2016-03-27    0.031092
2016-03-16    0.029610
2016-03-24    0.029342
2016-03-05    0.025327
2016-03-13    0.015670
2016-03-06    0.014043
2016-04-05    0.013096
2016-03-18    0.012911
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [31]:
#same values as above, but sorted by index
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The crawl takes place over 1 month, from looking at the minimum and maximum dates. It takes place in March and April 2016.

### Last Seen Column:

In [32]:
#looking at last_seen column proportions and sorted with index
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [33]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False)

2016-04-06    0.221806
2016-04-07    0.131947
2016-04-05    0.124761
2016-03-17    0.028086
2016-04-03    0.025203
2016-04-02    0.024915
2016-03-30    0.024771
2016-04-04    0.024483
2016-03-31    0.023783
2016-03-12    0.023783
2016-04-01    0.022794
2016-03-29    0.022341
2016-03-22    0.021373
2016-03-28    0.020859
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-23    0.018532
2016-03-26    0.016802
2016-03-16    0.016452
2016-03-15    0.015876
2016-03-19    0.015834
2016-03-27    0.015649
2016-03-14    0.012602
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007413
2016-03-18    0.007351
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64


The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

### Ad Created Column:


In [34]:
#look at column shape
autos["ad_created"].str[:10].unique().shape

(76,)

In [35]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()
        

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

### Registation year column

In [36]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

This column can show us the age of the car. Assuming it was registered soon after it was made, or bought. Oddly the minimum value of 1000 would be before cars were created, and the maximum of 9999 is the future which also doesn't make sense. We will have to deal with these outliers.

### Further analysis of registration year column

Realistically the maximum year should be 2016, which is when this dataset was created. Cars were invented around 1800, so lets assume 1900 for our minimum. 

In [37]:
#check we arent removing too much data
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

This is is around 4% of the data, so we will remove the entries:

In [38]:
autos[autos["registration_year"].between(1900,2016)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_T�V_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___T�V_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0       5000  control          bus               2004    manuell       158   
1       8500  control    limousine               1997  automatik       286   
2       8990     test    limousine               2009    manuell       102   
3       4350  control   kleinwagen               2007  automatik        71   
4       1350     test        kombi               2003    manuell         0   
...      ...      ...          ...                ...        ...       ...   
49995  24900  control    limousine               2011  automatik       239   
49996   1980  control       cabrio               1996    manuell        75   
49997  13200     test       cabrio               2014  automatik        69   
49998  22900  control        kombi               2013    manuell       150   
49999   1250  control    limousine               1996    manuell       101   

        model  odometer_km  registration_month fuel_type       brand  \
0      andere       150000                   3       lpg     peugeot   
1         7er       150000                   6    benzin         bmw   
2        golf        70000                   7    benzin  volkswagen   
3      fortwo        70000                   6    benzin       smart   
4       focus       150000                   7    benzin        ford   
...       ...          ...                 ...       ...         ...   
49995      q5       100000                   1    diesel        audi   
49996   astra       150000                   5    benzin        opel   
49997     500         5000                  11    benzin        fiat   
49998      a3        40000                  11    diesel        audi   
49999  vectra       150000                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[46681 rows x 17 columns]

In [39]:
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.064985
2005    0.060455
1999    0.059652
2004    0.055657
2003    0.055575
2006    0.054978
2001    0.054278
2002    0.051189
1998    0.048656
2007    0.046886
Name: registration_year, dtype: float64

Most cars were registered in the last 23 years (as of 2021)

## Exploring prices of brands

In [40]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.212828
opel              0.108658
bmw               0.108597
mercedes_benz     0.095789
audi              0.085823
ford              0.069639
renault           0.047874
peugeot           0.029445
fiat              0.025986
seat              0.018944
skoda             0.016061
nissan            0.015258
mazda             0.015217
smart             0.014290
citroen           0.014125
toyota            0.012581
hyundai           0.009945
sonstige_autos    0.009698
volvo             0.009039
mini              0.008607
mitsubishi        0.008216
honda             0.007989
kia               0.007104
alfa_romeo        0.006610
porsche           0.005910
suzuki            0.005889
chevrolet         0.005663
chrysler          0.003480
dacia             0.002656
daihatsu          0.002512
jeep              0.002224
subaru            0.002121
land_rover        0.002039
saab              0.001627
daewoo            0.001565
jaguar            0.001524
trabant           0.001400
r

In [41]:
autos["brand"].describe()

count          48565
unique            40
top       volkswagen
freq           10336
Name: brand, dtype: object

Volswagen is the most popular brand. Most of the German manufacters make up the top 10. We will analyse the top 10% of brands.

In [44]:
#creating new variable to store unique values of brand column
brand_count = autos["brand"].value_counts(normalize=True)

#looking at first 10 values
common_brands = brand_count[:10].index
print(common_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat'],
      dtype='object')


In [46]:
#create dictionary to store mean prices of each brand
mean_price_brands = {}

for brand in common_brands:
    brand_name = autos[autos["brand"] == brand ]
    mean_price = brand_name["price"].mean()
    mean_price_brands[brand] = int(mean_price)
    
mean_price_brands

{'volkswagen': 5332,
 'opel': 2944,
 'bmw': 8261,
 'mercedes_benz': 8536,
 'audi': 9212,
 'ford': 3728,
 'renault': 2431,
 'peugeot': 3065,
 'fiat': 2793,
 'seat': 4315}

Audi is the most expensive brand, followed by mercedes benz and bmw. The cheapest is fiat. Volswagen, the most popular brand, is mid priced which may explain its popularity.

## Mileage of the cars

Lets check if the average mileage has a link with mean price.

In [49]:
bmp_series = pd.Series(mean_price_brands)
pd.DataFrame(bmp_series, columns=["mean_price"])


mean_price
volkswagen           5332
opel                 2944
bmw                  8261
mercedes_benz        8536
audi                 9212
ford                 3728
renault              2431
peugeot              3065
fiat                 2793
seat                 4315

In [50]:
#creating dictionary to store mean mileage for each brand
brand_mean_mile = {}

for brand in common_brands:
    brand_name = autos[autos["brand"]== brand]
    mile_mean = brand_name["odometer_km"].mean()
    brand_mean_mile[brand] = int(mile_mean)
    
brand_mean_mile

{'volkswagen': 128896,
 'opel': 129383,
 'bmw': 132682,
 'mercedes_benz': 130796,
 'audi': 129492,
 'ford': 124349,
 'renault': 128062,
 'peugeot': 127356,
 'fiat': 117567,
 'seat': 121641}

In [51]:
mean_mileage = pd.Series(brand_mean_mile).sort_values(ascending=False)
mean_prices = pd.Series(mean_price_brands).sort_values(ascending=False)

In [52]:
#storing mean milage into a seperate index
mile_df = pd.DataFrame(mean_mileage, columns=['mean_mileage'])
mile_df

mean_mileage
bmw                  132682
mercedes_benz        130796
audi                 129492
opel                 129383
volkswagen           128896
renault              128062
peugeot              127356
ford                 124349
seat                 121641
fiat                 117567

In [53]:
mile_df["mean_price"] = mean_prices
mile_df

mean_mileage  mean_price
bmw                  132682        8261
mercedes_benz        130796        8536
audi                 129492        9212
opel                 129383        2944
volkswagen           128896        5332
renault              128062        2431
peugeot              127356        3065
ford                 124349        3728
seat                 121641        4315
fiat                 117567        2793

BMW's have the higest milage followed by Mercedes. These 2 brands are more expensive on average. In general we can see the higher the mileage, the higher the price.

# Conclusion


* Most of the vehicles were first registered in the past 20 years.
* Volkswagen is by far the most popular brand.
* There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.
